central/declarativeconfig/types/extractors.go

```diff
@@ -0,0 +1,59 @@
+package types
+
+import (
+	"github.com/gogo/protobuf/proto"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/errox"
+	"github.com/stackrox/rox/pkg/utils"
+)
+
+// IDExtractor extracts the ID from proto messages.
+type IDExtractor func(m proto.Message) string
+
+// NameExtractor extracts the ID from proto messages.
+type NameExtractor func(m proto.Message) string
+
+// UniversalIDExtractor provides a way to extract the ID from proto messages.
+func UniversalIDExtractor() IDExtractor {
+	return extractIDFromProtoMessage
+}
+
+// UniversalNameExtractor provides a way to extract the name from proto messages.
+func UniversalNameExtractor() NameExtractor {
+	return extractNameFromProtoMessage
+}
+
+func extractIDFromProtoMessage(message proto.Message) string {
+	// Special case, as the group specifies the ID nested within the groups properties.
+	if group, ok := message.(*storage.Group); ok {
+		return group.GetProps().GetId()
+	}
+	// Special case, as the name of the role is the ID.
+	if role, ok := message.(*storage.Role); ok {
+		return role.GetName()
+	}
+
+	messageWithID, ok := message.(interface {
+		GetId() string
+	})
+	// Theoretically, this should never happen unless we add more proto messages to the reconciliation. Hence, we use
+	// utils.Should to guard this.
+	if !ok {
+		utils.Should(errox.InvariantViolation.Newf("could not retrieve ID from message type %T %+v",
+			message, message))
+		return ""
+	}
+	return messageWithID.GetId()
+}
+
+func extractNameFromProtoMessage(message proto.Message) string {
+	messageWithName, ok := message.(interface {
```

Comment:
I guess we are back to good old serialized key for groups 🥴 

I'm not sure how much it's better, given that the auth provider ID will be non-telling for customers to identify the auth provider (imagine you have two groups sharing the same attributes / role name, but for different auth providers). Guess the same applies right now with the group ID, so I changed it to be the tuple of these values

Reply:
> I guess we are back to good old serialized key for groups 🥴

“The more you know about the past, the better prepared you are for the future.” —Theodore Roosevelt

> given that the auth provider ID will be non-telling for customers to identify the auth provider

I guess you're right, they should know what they are looking at. What do you think about terrifying 
`auth_provider_id:<id>:key:<key>:value:<value>:role_name:<roleName>` format? A more pleasant alternative would be `group <key>:<value>:<roleName> for auth provider ID <authProviderID>`

Reply:
Yeah I'm fine doing it either way, since we can be flexible with the name.
The root issue we have is that the group has no name and the auth provider is in ID-format instead of name, so looking at the declarative config it may be non-trivial identifying the groups.

Updated the names to the `group <key>:<value>:<roleName> for auth provider ID <authProviderID>` format.